_posts/2023-01-19-blog-new-svg-engine.md

```diff
@@ -0,0 +1,201 @@
+
+Due to the sheer capabilities of the Web platform, it has grown to an environment in which any kind of application can be developed -- ranging from video editing applications, big data processing pipelines to 3D games, all using JS/WebAssembly in a browser, presented using HTML5/CSS. And as an important bonus: in 2023, it's much easier to find and attract talented Web developers and designers that are fluent in HTML/CSS/JS, than those that are comfortable designing UI applications in proprietary, closed-source C/C++ frameworks.
+
+A long-term customer, successfully using WPE in their products, had very similar thoughts and carried out a study, contracting external Web designers to build a complete UI prototype using Web technology. The mock-up made extensive use of SVG2, embedded inline into HTML5 documents or via other mechanisms (CSS `background-image`, etc.). The UI fulfilled all expectations and worked great in Blink and WebKit-based browsers, delivering smooth animations. On the target device, however, the performance was too slow, far away from usable. A thorough analysis revealed that large parts of the Web page were constantly repainted, and layout operations were repeated for every frame when animations were active. The accumulated time to display a new frame during animations was in the order of a few milliseconds on desktop machines, but took 20-25 milliseconds on the target device, making smooth 60 FPS animations *impossible*.
+
+The poor performance is not the result of shortcomings in the WPE port of WebKit: when replacing the aforementioned animated SVG document fragments with HTML/CSS "equivalents" (e.g. simulating SVG circles with CSS `border-radius` tricks) the performance issue vanisheed. Why? SVG lacks support for a key feature called **accelerated compositing**, which has been available for HTML/CSS since its introduction more than a decade ago. This compositing heavily relies on the **Layer Tree**, which is unaware of SVG. Extending the **Layer Tree** implementation to account for SVG is the **main motivation for LBSE**.
+
+If you are unfamiliar with the concepts of **Render Tree** and **Layer Tree**, you might want to read the *"Key concepts"* section of an earlier <a href="https://blogs.igalia.com/nzimmermann/posts/2021-10-29-layer-based-svg-engine/page/3/#key-concepts">LBSE design document</a>, which provides an overview of the topic.
+
+### Prototyping
+
+The LBSE effort began in October 2019 as **a research project**, to find out an ideal design for the SVG **Render Tree**, that allows SVG to re-use the existing **Layer Tree** implementation (see **RenderLayer** / **RenderLayerCompositor** / **RenderLayerBacking**) with minimal changes. The aim for LBSE is to share as much code as possible with the HTML/CSS implementation, removing the need for things like SVG specific clipping/masking/filter code and disjoint HTML counterparts for the same operations.
```

Comment:
The parenthetical “(see **RenderLayer** / **RenderLayerCompositor** / **RenderLayerBacking**)” seems to call for the reader to see something else, but provides no link or other reference.  Does this need to be reworded, or have a link added?

Reply:
Indeed, good point. I think the whole paranthesis can go away - naming the classes doesn't add much value and the last sentence before is: "If you are unfamiliar with the concepts of **Render Tree** and **Layer Tree**, you might want to read the *"Key concepts"* section of an earlier <a href="https://blogs.igalia.com/nzimmermann/posts/2021-10-29-layer-based-svg-engine/page/3/#key-concepts">LBSE design document</a>, which provides an overview of the topic."

That is sufficient.

Reply:
Pushed a fix, removing the parenthesis.